BitwardenShared/UI/Auth/Utilities/VaultUnlockSetupHelper.swift

```diff
@@ -0,0 +1,149 @@
+}
+
+// MARK: - DefaultVaultUnlockSetupHelper
+
+/// A default implementation of `VaultUnlockSetupHelper` which is used to set up vault unlock methods.
+///
+@MainActor
+class DefaultVaultUnlockSetupHelper {
+    // MARK: Types
+
+    typealias Services = HasAuthRepository
+        & HasBiometricsRepository
+        & HasErrorReporter
+        & HasStateService
+
+    // MARK: Properties
+
+    /// The services used by this helper.
+    private let services: Services
+
+    // MARK: Initialization
+
+    /// Creates a new `DefaultVaultUnlockSetupHelper`.
+    ///
+    /// - Parameters:
+    ///   - services: The services used by this helper.
+    ///
+    init(services: Services) {
+        self.services = services
+    }
+
+    // MARK: Private
+
+    /// Shows an alert for the user to enter their pin for vault unlock.
+    ///
+    /// - Parameter showAlert: A closure used to handling showing the alert.
+    /// - Returns: The pin that the user entered.
+    ///
+    private func showEnterPinAlert(showAlert: @escaping (Alert) -> Void) async throws -> String {
+        try await withCheckedThrowingContinuation { continuation in
+            showAlert(.enterPINCode(
+                onCancelled: {
+                    continuation.resume(throwing: VaultUnlockSetupHelperError.userCancelled)
+                },
+                completion: { pin in
+                    continuation.resume(returning: pin)
+                }
+            ))
+        }
+    }
+
+    /// Shows an alert asking the user whether they want to require their master password to unlock
+    /// the vault when the app restarts.
+    ///
+    /// - Parameters:
+    ///   - biometricType: The biometric type the app supports.
+    ///   - showAlert: A closure used to handle showing the alert.
+    /// - Returns: Whether the user wants to require their master password on app restart.
+    ///
+    private func showUnlockWithPinAlert(
+        biometricType: BiometricAuthenticationType?,
+        showAlert: @escaping (Alert) -> Void
+    ) async -> Bool {
+        await withCheckedContinuation { continuation in
+            showAlert(.unlockWithPINCodeAlert(biometricType: biometricType) { requirePassword in
+                continuation.resume(returning: requirePassword)
+            })
+        }
+    }
+}
+
+// MARK: - DefaultVaultUnlockSetupHelper+VaultUnlockSetupHelper
+
+extension DefaultVaultUnlockSetupHelper: VaultUnlockSetupHelper {
+    func setBiometricUnlock(enabled: Bool, showAlert: @escaping (Alert) -> Void) async -> BiometricsUnlockStatus? {
+        do {
+            try await services.authRepository.allowBioMetricUnlock(enabled)
+            return try await services.biometricsRepository.getBiometricUnlockStatus()
+        } catch {
+            services.errorReporter.log(error: error)
+            showAlert(.defaultAlert(title: Localizations.anErrorHasOccurred))
+            return try? await services.biometricsRepository.getBiometricUnlockStatus()
+        }
```

Comment:
🤔 This line `return try? await services.biometricsRepository.getBiometricUnlockStatus()` is the same as the above one but with `try?` so why not `return nil` directly at the end of the catch? Can `getBiometricUnlockStatus()` change somehow in between? If so, could a comment be added how that situation may happen?

Reply:
This is a bit of an edge case. If `authRepository.allowBioMetricUnlock()` throws an error, I'd still want to return the current unlock status. Returning `nil` will ultimately remove the biometrics toggle from the screen, since we use that to determine whether biometrics are available or not. I can add a comment.